server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -223,4 +303,21 @@ public String toString() {
     public boolean isEmpty() {
         return shards.isEmpty();
     }
+
+    private boolean invariant() {
+
+        // initializingShards must consistent with that in shards
+        Collection<ShardRouting> shardRoutingsInitializing =
+            shards.values().stream().filter(ShardRouting::initializing).collect(Collectors.toList());
+        assert initializingShards.size() == shardRoutingsInitializing.size();
+        assert initializingShards.containsAll(shardRoutingsInitializing);
```

Comment:
This doesn't assert that the iteration orders are equal, but that is an important thing to assert.